src/Std/Data/DHashMap/Internal/List/Associative.lean

```diff
@@ -979,6 +1200,68 @@ theorem insertEntry_of_containsKey_eq_false [BEq α] {l : List ((a : α) × β a
     (h : containsKey k l = false) : insertEntry k v l = ⟨k, v⟩ :: l := by
   simp [insertEntry, h]
 
+theorem insertEntry_of_containsKey_eq_false_perm [BEq α] {l : List ((a : α) × β a)} {k : α} {v : β k}
+    (h : containsKey k l = false) : Perm (insertEntry k v l) (⟨k,v⟩::l) := by
+  cases l with
+  | nil => simp
+  | cons hd tl => simp[insertEntry, h]
+
+theorem insertEntry_mem_of_different_keys [BEq α] [LawfulBEq α] {l : List ((a : α) × β a)} {k : α} {v : β k} {a: ((a : α) × β a)} {h: ¬ ⟨k,v⟩ = a} {distinct: List.Pairwise (fun a b => ! (a.1 == b.1) ) l}: ⟨k, v⟩ ∈ insertEntry a.fst a.snd l ↔ ⟨k,v⟩ ∈ l ∧ (k == a.fst) = false := by
```

Comment:
As I'm sure you've noticed these proofs characterizing membership are long and complicated. But in fact you should be able to do without them. You need these two pieces of infrastructure:
```lean
theorem DistinctKeys.insertList [BEq α] [PartialEquivBEq α] {l₁ l₂ : List ((a : α) × β a)} (h : DistinctKeys l₁) :
    DistinctKeys (insertList l₁ l₂) := by
  induction l₂ using assoc_induction generalizing l₁
  · simpa [insertList]
  · rename_i k v t ih
    rw [insertList.eq_def]
    exact ih h.insertEntry

theorem insertList_cons_perm [BEq α] [EquivBEq α] (l₁ l₂ : List ((a : α) × β a))
    (p : (a : α) × β a) (hl₁ : DistinctKeys l₁) (hl₂ : DistinctKeys (p :: l₂)) :
    (insertList l₁ (p :: l₂)).Perm (insertEntry p.1 p.2 (insertList l₁ l₂)) := by
  induction l₂ generalizing p l₁
  · simp [insertList]
  · rename_i h t ih
    rw [insertList]
    refine (ih _ _ hl₁.insertEntry hl₂.tail).trans
      ((insertEntry_of_perm hl₁.insertList
        (ih _ _ hl₁ (distinctKeys_of_sublist (by simp) hl₂))).trans
      (List.Perm.trans ?_ (insertEntry_of_perm hl₁.insertList.insertEntry
        (ih _ _ hl₁ (distinctKeys_of_sublist (by simp) hl₂)).symm)))
    apply getEntry?_ext hl₁.insertList.insertEntry.insertEntry
      hl₁.insertList.insertEntry.insertEntry (fun k => ?_)
    simp only [getEntry?_insertEntry]
    split <;> rename_i hp <;> split <;> rename_i hh <;> try rfl
    rw [DistinctKeys.def] at hl₂
    have := List.rel_of_pairwise_cons hl₂ (List.mem_cons_self _ _)
    simp [BEq.trans hh (BEq.symm hp)] at this
```

What this lemma essentially does is commute `insertList` and `insertEntry` past each other.

This will allow you do to "normal" induction proofs about `insertList`. As an example, the proof of `getValueCast?_insertList_toInsert_mem` is straightforward using this approach:

```lean
theorem getValueCast?_insertList_toInsert_mem [BEq α] [LawfulBEq α] (l toInsert: List ((a : α) × β a)) {k k': α} {k_eq: k == k'} {v: β k}
    {distinct: List.Pairwise (fun a b => (a.1 == b.1) = false) toInsert} {distinct2: DistinctKeys l} {mem: ⟨k,v⟩ ∈ toInsert}: getValueCast? k' (insertList l toInsert) = some (cast (by congr;apply LawfulBEq.eq_of_beq k_eq) v) := by
  rw [← DistinctKeys.def] at distinct
  induction mem
  · rename_i l'
    rw [getValueCast?_of_perm distinct2.insertList (insertList_cons_perm _ _ _ distinct2 distinct)]
    simp [k_eq, getValueCast?_insertEntry]
  · rename_i p l' hl' ih
    rw [getValueCast?_of_perm distinct2.insertList (insertList_cons_perm _ _ _ distinct2 distinct)]
    rw [getValueCast?_insertEntry, dif_neg, ih distinct.tail]
    have := distinct.containsKey_eq_false
    rw [containsKey_eq_false_iff] at this
    simpa using BEq.neq_of_neq_of_beq (this _ hl') k_eq
```

Using this approach, you should hopefully be able to drastically reduce the total length of the proofs in this file.

Reply:
This makes things way easier indeed. Thank you! :)